bin/prepare_score_distribution.py

```diff
@@ -86,46 +131,90 @@ def main(args=None):
     elif not os.path.exists(args.outdir):
         os.makedirs(args.outdir)
 
-    print("Joining input data...", file=sys.stderr, flush=True, end="")
+    print("Prepare df with protein info ...", flush=True)
+    # Prepare df for joining protein information
+    # (get condition_name, entity_weights and filter for condition entities)
+    protein_info = (
+        microbiomes_entities_occs.merge(conditions)
+        .drop(columns="microbiome_id")
+        .merge(condition_allele_map)
+        .drop(columns="condition_id")
+        .merge(entities_proteins_occs)
+        .drop(columns="entity_id")
+    )
+    # -> protein_id, entity_weight, condition_name, allele_id
+    # merged against condition_allele_map to keep only entities, and thus proteins, for which a prediction is requested for the current allele
+    print("\nInfo: protein_info", flush=True)
+    protein_info.info(verbose=False, memory_usage=print_mem)
 
-    # for each allele separately (to save mem)
+    # Prepare output files for each allele
+    outfile_dict = {}
     for allele_id in alleles.allele_id:
-        print("Process allele: ", allele_id, flush=True)
-
-        data = (
-            predictions[predictions.allele_id == allele_id]
-            .merge(protein_peptide_occs)
-            .merge(entities_proteins_occs)
-            .drop(columns="protein_id")
-            .merge(microbiomes_entities_occs)
-            .drop(columns="entity_id")
-            .merge(conditions)
-            .merge(condition_allele_map)
-            .drop(columns=["allele_id", "microbiome_id", "condition_id"])
-            .groupby(["peptide_id", "prediction_score", "condition_name"])["entity_weight"]
-            .sum()
-            .reset_index(name="weight_sum")
-            .drop(columns="peptide_id")
-        )
-        # NOTE
-        # for each peptide in a condition the weight is computed as follows:
-        # - the sum of all weights of the corresponding entity_weights, each weighted by the number of proteins in which the peptide occurs
-        # - multiple occurrences of the peptide within one protein are not counted
-
-        with open(os.path.join(args.outdir, "prediction_scores.allele_" + str(allele_id) + ".tsv"), "w") as outfile:
-            data[["prediction_score", "condition_name", "weight_sum"]].to_csv(
-                outfile, sep="\t", index=False, header=True
+        outfile = open(os.path.join(args.outdir, "prediction_scores.allele_" + str(allele_id) + ".tsv"), "w")
+        outfile_dict[allele_id] = outfile
+    print_header = True
+
+    try:
+        # Process predictions chunk-wise based on peptide_ids
+        # (predictions chunk can contain more than chunk_size rows due to multiple alleles)
+        max_peptide_id = predictions.index.max()
+        for i in range(0, max_peptide_id, args.chunk_size):
```

Comment:
Will this actually use the last chunk? Afaik `range(2, 10, 3)` only return [2, 5, 8], but not 10. Maybe here it would make sense to use a function e.g. `process_predictions(pep_id_start, pep_id_stop, ...)` and generate the start and stop tuples beforehand. Another solution would be to generate them instead of the `range(...)` using something like this:
```
max_peptide_id = 10
chunk_size = 3
for i in [(ch_start, ch_start+ chunk_size-1) if ch_start+ chunk_size-1 <= end else (start, max_peptide_id) for ch_start in range(0, max_peptide_id, chunk_size)]:
    print(i)
```

While writing this comment I saw, that you are collecting the peptide_ids via the chunk size afterwards, so this comment is not necessary. Still maybe a suggestion for better readability?